projects.go

```diff
@@ -1287,57 +1283,55 @@ type ProjectFile struct {
 	Markdown string `json:"markdown"`
 }
 
-// UploadFile upload a file from disk
+// UploadFile uploads a file.
 //
 // GitLab API docs: https://docs.gitlab.com/ce/api/projects.html#upload-a-file
-func (s *ProjectsService) UploadFile(pid interface{}, file string, options ...RequestOptionFunc) (*ProjectFile, *Response, error) {
+func (s *ProjectsService) UploadFile(pid interface{}, content io.Reader, filename string, options ...RequestOptionFunc) (*ProjectFile, *Response, error) {
 	project, err := parseID(pid)
 	if err != nil {
 		return nil, nil, err
 	}
 	u := fmt.Sprintf("projects/%s/uploads", PathEscape(project))
 
-	f, err := os.Open(file)
+	options = append(options, WithFile(content, filename, UploadFile))
+	req, err := s.client.NewRequest(http.MethodPost, u, nil, options)
 	if err != nil {
 		return nil, nil, err
 	}
-	defer f.Close()
-
-	b := &bytes.Buffer{}
-	w := multipart.NewWriter(b)
 
-	_, filename := filepath.Split(file)
-	fw, err := w.CreateFormFile("file", filename)
+	pf := new(ProjectFile)
+	resp, err := s.client.Do(req, pf)
 	if err != nil {
-		return nil, nil, err
+		return nil, resp, err
 	}
 
-	_, err = io.Copy(fw, f)
-	if err != nil {
-		return nil, nil, err
-	}
-	w.Close()
+	return pf, resp, nil
+}
 
-	req, err := s.client.NewRequest(http.MethodPost, u, nil, options)
+// UploadAvatar uploads an avatar.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/projects.html#upload-a-project-avatar
+func (s *ProjectsService) UploadAvatar(pid interface{}, avatar io.Reader, options ...RequestOptionFunc) (*Project, *Response, error) {
+	project, err := parseID(pid)
 	if err != nil {
 		return nil, nil, err
 	}
+	u := fmt.Sprintf("projects/%s", PathEscape(project))
 
-	// Set the buffer as the request body.
-	if err = req.SetBody(b); err != nil {
+	options = append(options, WithFile(avatar, "avatar.png", UploadAvatar))
```

Comment:
I do not think this works. You assume avatars are just `.png` files here. You should allow `filename` argument.